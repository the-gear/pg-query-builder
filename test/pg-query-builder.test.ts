import SelectBuilder from '../src/pg-query-builder';
import './serializer';

/**
 * SelectBuilder
 */
describe('SelectBuilder', () => {
  it('is instantiable', () => {
    expect(new SelectBuilder()).toBeInstanceOf(SelectBuilder);
  });

  it('should create minimal query', () => {
    const query = new SelectBuilder().select('1 + 1 as result');
    expect(query).toMatchSnapshot();
  });

  it('should create simple query', () => {
    const query = new SelectBuilder('"testTable"')
      .select('a, b')
      .select('"c"')
      .select('1 + 1 as calculated');
    expect(query).toMatchSnapshot();
  });

  it('should throw if argument length mismatch', () => {
    expect(() => {
      new SelectBuilder('"testTable"').where('a = $$', 1, 2);
    }).toThrow();
  });

  it('can change command', () => {
    const query = new SelectBuilder();
    query
      .setCommand('SELECT DISTINCT')
      .select('a')
      .select('b')
      .from('TEST');
    expect(query).toMatchSnapshot();
  });

  it('should create complex query', () => {
    expect(createComplexQuery({})).toMatchSnapshot();

    expect(
      createComplexQuery({
        userId: '123',
        fromDate: new Date(1542673386000),
        endDate: null,
        kinds: ['IN', 'OUT'],
        peer: 'the-peer',
        limit: 20,
      }),
    ).toMatchSnapshot();
  });
});

function createComplexQuery({
  userId,
  limit,
  fromDate,
  endDate,
  kinds,
  peer,
}: any) {
  const query = new SelectBuilder('"Table"');

  query
    .select(
      `CONCAT('Expr:', "tableId") as "id"`,
      '"fromPerson"."personPublicId" as "from"',
      '"toPerson"."personPublicId" as "to",*',
    )
    .join(
      'JOIN "Person" "fromPerson" on "Table"."fromPersonId" = "fromPerson"."personId"',
    )
    .join(
      'JOIN "Person" "toPerson" on "Table"."toPersonId" = "toPerson"."personId"',
    )
    .setLimit(limit);

  if (userId) {
    query.where(
      '"fromPerson"."userId" = $$ OR "toPerson"."userId" = $$',
      userId,
      userId,
    );
  } else {
    query.where('FALSE');
  }

  if (fromDate) {
    query.where('"createdAt" >= $$', fromDate);
  }

  if (endDate) {
    query.where('"createdAt" <= $$', endDate);
  }

  if (kinds) {
    query.where('"kinds" IN ($$)', kinds);
  }

  if (peer) {
    query.where(
      '"fromPerson"."personPublicId" = $$ OR "toPerson"."personPublicId" = $$',
      peer,
      peer,
    );
  }

  query.orderBy('"createdAt" DESC');

  return query;
}
