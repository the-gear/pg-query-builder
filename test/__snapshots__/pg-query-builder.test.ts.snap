// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBuilder can change command 1`] = `
SQL {
  SELECT DISTINCT
  a, b
  FROM TEST
} Arguments Array []
`;

exports[`SelectBuilder should create complex query 1`] = `
SQL {
  SELECT
  CONCAT('Expr:', "tableId") as "id", "fromPerson"."personPublicId" as "from", "toPerson"."personPublicId" as "to",*
  FROM "Table"
  JOIN "Person" "fromPerson" on "Table"."fromPersonId" = "fromPerson"."personId"
  JOIN "Person" "toPerson" on "Table"."toPersonId" = "toPerson"."personId"
  WHERE (FALSE)
  ORDER BY "createdAt" DESC
} Arguments Array []
`;

exports[`SelectBuilder should create complex query 2`] = `
SQL {
  SELECT
  CONCAT('Expr:', "tableId") as "id", "fromPerson"."personPublicId" as "from", "toPerson"."personPublicId" as "to",*
  FROM "Table"
  JOIN "Person" "fromPerson" on "Table"."fromPersonId" = "fromPerson"."personId"
  JOIN "Person" "toPerson" on "Table"."toPersonId" = "toPerson"."personId"
  WHERE ("fromPerson"."userId" = $1 OR "toPerson"."userId" = $2)
    AND ("createdAt" >= $3)
    AND ("kinds" IN ($4))
    AND ("fromPerson"."personPublicId" = $5 OR "toPerson"."personPublicId" = $6)
  ORDER BY "createdAt" DESC
  LIMIT 20
} Arguments Array [
  "123",
  "123",
  2018-11-20T00:23:06.000Z,
  Array [
    "IN",
    "OUT",
  ],
  "the-peer",
  "the-peer",
]
`;

exports[`SelectBuilder should create minimal query 1`] = `
SQL {
  SELECT
  1 + 1 as result
} Arguments Array []
`;

exports[`SelectBuilder should create simple query 1`] = `
SQL {
  SELECT
  a, b, "c", 1 + 1 as calculated
  FROM "testTable"
} Arguments Array []
`;

exports[`SelectBuilder should query all columns if none explicitly selected 1`] = `
SQL {
  SELECT
  *
  FROM test
} Arguments Array []
`;
